### ISSUE TITLE
Implement Markerclusterer in GWT Google Maps V3 ,through JSNI

### ISSUE THREAD

**Issue by MarioMartin** [branflake2267/GWT-Maps-V3-Api#20]
could help me please: 
I have a GWT project with GWT GoogleMapsAPI v3 and would like to implement Markerclusterer through a JSNI method.

My question is how to manipulate the MapWidget within JSNI method

 this is my code but does not work.

package cl.prueba.JS.client;

``` java
import com.google.gwt.core.client.EntryPoint;
import com.google.gwt.dom.client.Element;
import com.google.gwt.maps.client.MapOptions;
import com.google.gwt.maps.client.MapTypeId;
import com.google.gwt.maps.client.MapWidget;
import com.google.gwt.maps.client.base.LatLng;
import com.google.gwt.user.client.ui.RootPanel;

/**
 * Entry point classes define <code>onModuleLoad()</code>.
 */
public class PruebaJS implements EntryPoint {

final private static int    ZOOM     = 3;
final private static LatLng CENTER   = new LatLng(-33.437912, -70.650535);
final private static String MAP_TYPE = new MapTypeId().getRoadmap();
private MapWidget           mapa ;


private void inicializaMapa() {
    final MapOptions options = new MapOptions();
    options.setZoom(ZOOM);
    options.setCenter(CENTER);
    options.setMapTypeId(MAP_TYPE);
    options.setDraggable(true);
    options.setNavigationControl(true);
    options.setMapTypeControl(true);
    mapa = new MapWidget(options);
    mapa.setSize("300px","300px");

                    iniMarkerClusterer(mapa);

    }

public  native MapWidget  iniMarkerCluster(MapWidget map ) /*-{

    var markers = [];
    for ( var i = 0; i < 100; i++) {
    var dataPhoto = $wnd.data.photos[i];
    var latLng = new $wnd.google.maps.LatLng(dataPhoto .latitude,dataPhoto.longitude );      
    var marker = new $wnd.google.maps.Marker({position : latLng});
    markers.push(marker);
    }
    var markerCluster = new $wnd.MarkerClusterer(map , markers);

    }-*/;


    public void onModuleLoad() {

        RootPanel.get("mapa").add(mapa));

    }

}
/**********************************************************************************************************************///
```

my html is :
HTML ....

``` html
<script type="text/javascript" src="pruebajs/pruebajs.nocache.js"></script>
<script type="text/javascript"src="http://maps.google.com/maps/api/js?sensor=false"></script>
<script type="text/javascript" src="pruebajs/data.json"></script>
<script type="text/javascript" src="pruebajs/markerclusterer.js"></script>

<body>
    <div id="mapa"></div>
</body>
```

Thank you very much!!
I hope they understand my problem, if they have questions, please let me know.

atte 
Mario

**Comment by branflake2267** [comment]
http://code.google.com/p/google-maps-utility-library-v3/wiki/Libraries - Someone asked if we could implement the cluster

http://code.google.com/p/google-maps-utility-library-v3/source/browse/trunk/markerclusterer/docs/reference.html

**Comment by branflake2267** [comment]
This library is huge, but does have some inviting utility classes. Maybe after 3.10 I might endeavor on.

**Comment by twistedpair** [comment]
Good point. From a low ever GWT perspective, I'm sometimes weary of large bolt on libraries. For example the Garmin Communicator API downloads nearly 2 dozen js libs and jquery to do even the simplest task. Hence I did gwt-garmin-api as a native implementation in just 8k.

Hence, to include something like that, we'd need to be sure to make it optional in the loader.

**Comment by branflake2267** [comment]
Sounds like a nifty api. :)

I agree. I was thinking separate module, and maybe child modules with in that.

**Comment by twistedpair** [comment]
Correct. What I really love about GWT is the minimalism of the final compiled product while many a framework these days bolts dependencies on top of dependencies. I recall a major retailer's landing page that required 276 downloads to load.

**Comment by branflake2267** [comment]
Great point! ... Holy moly 276...

**Comment by vhmeirelles** [comment]
Hi,

Does anyone have an example of using the MarkerClusterer in GWT Maps API V3?
Thanks a lot.

**Comment by branflake2267** [comment]
I need to investigate to see if this fits this issue:
https://github.com/branflake2267/GWT-Maps-V3-Api/tree/master/gwt-maps-utility/markerclustererplus

**Comment by sherloclops** [comment]
Did you ever do this? I tried your MarkerClusterer and I get a "operation not supported error" by javascript... this after custom loading the compiled js implementing the object and trying to customize your code.

**Comment by sherloclops** [comment]
http://google-maps-utility-library-v3.googlecode.com/svn/trunk/markerclusterer/docs/examples.html

Seems to work fine especially with the custom js... not sure where I see that js library in your utilities modules either since MarkerClusterer isn't part of the v3 library.

Maps become useless once you have a few hundreds markers so this is a critical class... in spite of it being on the fringe of maps v3.

**Comment by branflake2267** [comment]
Hm, I don't see a javascript loader. Looks like some work still needs to be done yet. Sorry I haven't had time to investigate further since this is a volunteer project I've had some time conflicts. Patches are welcome.

**Comment by sherloclops** [comment]
I added:

``` xml
    <inherits name='com.google.gwt.maps.Maps' />
    <inherits name='com.google.gwt.maps.utility.markerclustererplus.MarkerClustererPlus' />
    <script src="http://maps.google.com/maps/api/js?sensor=false"></script>
```

to my gwt.xml

also also

``` xml
        <dependency>
            <groupId>com.github.branflake2267</groupId>
            <artifactId>gwt-maps-api</artifactId>
            <version>3.10.0-alpha-6</version>
        </dependency>
        <dependency>
            <groupId>com.github.branflake2267</groupId>
            <artifactId>gwt-maps-utility-markerclustererplus</artifactId>
            <version>3.10.0-alpha-6</version>
        </dependency>
```

In my POM.

So far so good.

But when I try:

``` java
        final MapWidget map = new MapWidget(mapOptions);

        final MarkerClustererOptions clusterOptions = MarkerClustererOptions.newInstance(); 
        clusterOptions.setAverageCenter( true );

        manager = MarkerClusterer.newInstance( map, clusterOptions );
```

It blow up here with:

``` bash
18:51:22.469 [ERROR] [gwt] Uncaught exception escaped
com.google.gwt.core.client.JavaScriptException: (TypeError) @com.google.gwt.maps.utility.markerclustererplus.client.MarkerClusterer::createJso(Lcom/google/gwt/core/client/JavaScriptObject;Lcom/google/gwt/core/client/JavaScriptObject;Lcom/google/gwt/core/client/JavaScriptObject;)([JavaScript object(1517), JavaScript object(1519), JavaScript object(1518)]): $wnd.MarkerClusterer is not a constructor
```

Now I took the 

``` xml
    <script src="http://maps.google.com/maps/api/js?sensor=false"></script>
```

from the html examples I saw successfully using the api.

But obviously something is amiss here. Hope your insight can resolve this in a few minutes :)

**Comment by sherloclops** [comment]
I see

``` java
  private static native MarkerClusterer createJso(JavaScriptObject map, JavaScriptObject markers, JavaScriptObject options) /*-{
      return new $wnd.MarkerClusterer(map, markers, options);
  }-*/;
```

which should work given the javascript example does pretty much the same thing.

**Comment by sherloclops** [comment]
I got it to work!!!!

Here's the recipe:

3.10.0-alpha-6 is fine.

Now in gwt.xml:

<script src="markerclusterer.js" />

create a public directory where the xml is and stick there a markerclusterer.js file ripped from:

http://code.google.com/p/google-maps-utility-library-v3/source/browse/trunk/markerclustererplus/src/markerclusterer.js?r=360

Then all will work!

Took me a bit to figure out but hope I made my little contribution to the community here.

**Comment by sherloclops** [comment]
also in gwt.xml add:

<script src="markerclusterer.js" />

And life will be good.

**Comment by branflake2267** [comment]
Good job.

**Comment by nasamad** [comment]
Hello sherloclops, I'm facing a similar issue: I added the markercluster.js in the same directory as my custom .gwt.xml
and in my gwt.xml, I added: [script src="markerclusterer.js" /]
But I'm still getting the same error. In which directory did you the JS file and how are you referencing it? Thanks

**Comment by branflake2267** [comment]
Try adding a timer and see if its loaded after a second.

**Comment by branflake2267** [comment]
I use a callback method for the maps api, and I don't see one here. If it has a the ajax loader strategy could be used so that a runnable could be used to programmatically load after the js has been injected.

**Comment by nasamad** [comment]
I created com.google.gwt.maps.utility.markerclustererplus package , copied the MarkerClusterPlus.gwt.xml into it, modified it to include  [script src="markerclusterer.js" /]. I then created the "public" directory, added the markermanager.js inside it.. It doesn't work in Development mode. But it does when I deploy it to the server.  "production".

**Comment by nasamad** [comment]
"I use a callback method for the maps api" Are you referring to  LoadApi.go(onLoad, loadLibraries, sensor);

**Comment by branflake2267** [comment]
The question is does the markermanager.js call a method once its loaded? If it does that can be used as a callback in a GWT runner, like the ajax loader method, which is similiar to the LoadApi.go... Some js has a call back feature, or feature that can be used for onsuccess or onfailure. Working with js has to be done asyncly, that is there is a timing issue between the time it gets loaded, parsed and injected into the dom. Once its in the dom, then its available to call using jsni methods. If calling a method to it is undefined, I suspect that it hasn't loaded by the time the js or jsni calls the method.

**Comment by nasamad** [comment]
Thanks for the explanation. The markermanager does trigger a loaded event once it's loaded. However, I don't see anywhere where such event is triggered  in the markercluster.js.

**Comment by branflake2267** [comment]
Cool, I've been meaning to look deeper at it but haven't had time.

**Comment by nasamad** [comment]
Brandon, I got it to work in dev mode by delaying the instantiation until MarkerCluster is actually needed. However, the events that I had registered to the markers themselves no longer work. Any trick that I can use to register mouseover and mouse events using this MarkerClusterer JSNI as shown here http://google-maps-utility-library-v3.googlecode.com/svn/trunk/markerclustererplus/examples/events_example.html ?  Thanks

**Comment by branflake2267** [comment]
I suspect the handlers have some sort of handling for events, and if so I would check out the event system. Its not pretty b/c there is some generic challenges in it and I used inheritance a bunch. Another thing that I do for widgets is use > https://github.com/branflake2267/GWT-Maps-V3-Api/blob/master/gwt-maps-api/src/main/java/com/google/gwt/maps/client/events/MapPanel.java to wire up the events but have to be careful b/c it can leak if not managed properly.

**Comment by branflake2267** [comment]
https://github.com/branflake2267/GWT-Maps-V3-Api/blob/master/gwt-maps-api/src/main/java/com/google/gwt/maps/client/events/MapHandlerRegistration.java contains the main part of the map event logic, which uses handlers to pass back objects

**Comment by nasamad** [comment]
I will give it a try. Thanks

**Comment by HarisHashim** [comment]
Hello & good day! Thanks for such awesome gwt lib for map.

1. Do I still have to do workaround as done by  nasamad? Or markercluster has been fixed in 3.10.0-alpha-7?

> I created com.google.gwt.maps.utility.markerclustererplus package , copied the MarkerClusterPlus.gwt.xml into it, modified it to include [script src="markerclusterer.js" /]. I then created the "public" directory, added the markermanager.js inside it.. It doesn't work in Development mode. But it does when I deploy it to the server. "production".

2. I did not do workaround done by nasamad and get this error "09:23:35 SEVERE [LogConfiguration] (TypeError) : $wnd.MarkerClusterer is not a constructor"

My code is a mess :D but I expect it to work. Above error is at the MarkerClusterer.newInstance line.

        private void loadMapApi() {
		boolean sensor = false;

		// load all the libs for use in the maps
		ArrayList<LoadLibrary> loadLibraries = new ArrayList<LoadApi.LoadLibrary>();
		loadLibraries.add(LoadLibrary.DRAWING);
		loadLibraries.add(LoadLibrary.GEOMETRY);
		loadLibraries.add(LoadLibrary.VISUALIZATION);

		infoWindowOptions = InfoWindowOptions.newInstance();

		Runnable onLoad = new Runnable() {
			@Override
			public void run() {
				drawMap();
			}
		};

		LoadApi.go(onLoad, loadLibraries, sensor, "key=keysabcdtoottoot");
        }

        private void drawMap() {

		if (this.iterator().hasNext()) {
			logger.debug("Not calling drawmap since Panel already contain a child!");
			return;
		} else {
			logger.debug("Calling drawmap since Panel does not contain a child!");
		}

		LatLng centerCoords = LatLng.newInstance(3.1389391564497657d, 101.6839599609375d);
		MapOptions opts = MapOptions.newInstance();
		opts.setZoom(14);
		opts.setCenter(centerCoords);
		opts.setMapTypeId(MapTypeId.ROADMAP);

		mapWidget = new MapWidget(opts);
		infoWindow = InfoWindow.newInstance(infoWindowOptions);

		final MarkerClustererOptions clusterOptions = MarkerClustererOptions.newInstance();
		clusterOptions.setAverageCenter(true);
		markerCluster = MarkerClusterer.newInstance(mapWidget, clusterOptions);

		if (mapWidget != null) {
			ov = OverlayView.newInstance(mapWidget, new ovOnDrawHandler(), new ovOnAddHandler(),
					new ovViewOnRemoveHandler());

			logger.debug("Adding click handler for map!");
			mapWidget.addClickHandler(new MapClickHandler());
			mapWidget.addRightClickHandler(new MapRightClickHandler());
		}

		logger.debug("Adding map widget to MapDashWidget");
		// this.clear();
		this.add(mapWidget);

		resize();
        }
